### ISSUE TITLE
redundant_pub_crate Vs unreachable-pub

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#8005]
I am not sure if this is a false positive. On this code:

```rust
#![allow(unused_variables, dead_code)]

#![warn(clippy::nursery)]

mod foos {
    pub(crate) struct Bar { x: usize }

    impl Bar {
        pub(crate) const fn new() -> Self {
            Self { x: 0 }
        }
    }
}

use foos::Bar;

fn main() {
    let b = Bar::new();
}
```

Clippy gives me:

```
    Checking foo1 v0.1.0 (C:\lavoro\metti_a_posto\progettini_in_corso\eulers\many_solutions\foo1)
warning: pub(crate) struct inside private module
 --> src\main.rs:6:5
  |
6 |     pub(crate) struct Bar { x: usize }
  |     ----------^^^^^^^^^^^
  |     |
  |     help: consider using: `pub`
  |
note: the lint level is defined here
 --> src\main.rs:3:9
  |
3 | #![warn(clippy::nursery)]
  |         ^^^^^^^^^^^^^^^
  = note: `#[warn(clippy::redundant_pub_crate)]` implied by `#[warn(clippy::nursery)]`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_pub_crate
```

But if I replace it with just "pub" and I compile the code with rustc using "-W unreachable-pub", I get:

```
warning: unreachable `pub` item
 --> ...\main.rs:6:5
  |
6 |     pub struct Bar { x: usize }
  |     ---^^^^^^^^^^^
  |     |
  |     help: consider restricting its visibility: `pub(crate)`
  |
  = note: requested on the command line with `-W unreachable-pub`
  = help: or consider exporting it for use by other crates
```

So I am not sure who is wrong here.

**Comment by pmnoxx** [comment]
I think `redundtant_pub_crate` is a bad lint, which should be removed. 


Consider following use case:

```rust
#![allow(unused_variables, dead_code)]

#![warn(clippy::nursery)]

mod foos {
    pub(crate) struct Bar { x: usize }

    pub struct Foo { x: usize }
}

use foos:Bar 

pub use foos::Bar;
```

Changing `pub(crate) ` to `pub`, works, the type is not visible outside. However, I wouldn't do it for a few reasons:
1) If both types were `pub`, I wouldn't know which is exported public-ally, therefore, which one needs to be documented.
2) If both types were `pub` and I wanted to refactor the code to make `foos`  `pub mod`. Then, It would be easy to make a mistake, and keep `pub` next to `struct Bar`, therefore exporting a private type, due to simple refactoring issue.

I think that the correct thing to do is to use lints which downgrade `pub` to `pub(crate)` whenever possible. Here is a good real life example: https://github.com/alexcrichton/toml-rs/blob/master/src/tokens.rs
Types are marked as `pub`, but  the `mod` is not. I wanted to use those types for another crate I'm creating, but they turned out to be private, which was not possible to tell by reading this file a alone.

**Comment by strohel** [comment]
I think is issue is a duplicate of #5369.

**Comment by leonardo-m** [comment]
Closed in favor of 5369 then.
